### PR TITLE
Add Live TestMode and Skip Recording Concepts

### PR DESCRIPTION
**Pull request by alzimmermsft**
Implements #5174 

This PR introduces two new concepts to the Azure Core testing framework.

A new value `LIVE` was added to the `TestMode` enum. This state is used to indicate that tests should be ran against a live service without recording the network calls, this is meant to help reduce the overhead of testing by removing the recording process. The use case for this change is to remove recording tests during the nightly validation and in cases where sensitive information that hasn't been masked is being sent during testing.

A new test annotation, `DoNotRecord`, was added. This annotation is meant to be used with tests that don't want their network calls record or they make no network calls but surrounding testing infrastructure does. This annotation has an optional parameter of `skipInPlayback` which is meant to be used with tests that make network calls but won't be recorded due to the inability to redact sensitive data or they call into code paths that cannot be mocked.

### CODE REVIEW

File: sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationClientTestBase.java [hunk 1]
```diff
@@ -34,46 +30,48 @@
...
-        }
+        clientBuilder = new ConfigurationClientBuilder()
+            .connectionString(connectionString)
+            .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS));
 
-        return Objects.requireNonNull(client);
+        if (testMode == TestMode.PLAYBACK && !testRunVerifier.doNotRecordTest()) {
+            clientBuilder.httpClient(interceptorManager.getPlaybackClient());
+        } else if (testMode == TestMode.RECORD && !testRunVerifier.doNotRecordTest()) {
+            clientBuilder.httpClient(new NettyAsyncHttpClientBuilder().wiretap(true).build())
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationClientTestBase.java]: `LIVE` won't need Netty wiretap set to true as it doesn't need to inspect the request or response for recording purposes.

File: sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationClientTestBase.java [hunk 2]
```diff
@@ -34,46 +30,48 @@
...
-            client = null;
-        }
+        clientBuilder = new ConfigurationClientBuilder()
+            .connectionString(connectionString)
+            .httpLogOptions(new HttpLogOptions().setLogLevel(HttpLogDetailLevel.BODY_AND_HEADERS));
 
-        return Objects.requireNonNull(client);
+        if (testMode == TestMode.PLAYBACK && !testRunVerifier.doNotRecordTest()) {
+            clientBuilder.httpClient(interceptorManager.getPlaybackClient());
+        } else if (testMode == TestMode.RECORD && !testRunVerifier.doNotRecordTest()) {
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/appconfiguration/azure-data-appconfiguration/src/test/java/com/azure/data/appconfiguration/ConfigurationClientTestBase.java]: `LIVE` and `DoNotRecord` serve slightly different but related purposes. `LIVE` won't record anything but the `DoNotRecord` annotation is meant for tests that shouldn't have any data recorded.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/DoNotRecord.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * Annotation given to some tests to indicate that network calls made during the test shouldn't be recorded.
+ *
+ * <p>
+ * Pass {@code true} for {@link #skipInPlayback() skipInPlayback} to indicate that the test shouldn't run when tests are
+ * ran in {@link TestMode#PLAYBACK}. A common case for setting this to {@code true} is when the test has either
+ * sensitive content that cannot be redacted or calls into code that cannot be mocked.
+ */
+@Retention(RUNTIME)
+@Target({METHOD})
+public @interface DoNotRecord {
```
Comment (srnagar) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/DoNotRecord.java]: Just for my understanding, why would any test run only in live mode but shouldn't be recorded?

Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/DoNotRecord.java]: Tests that contain information that cannot be recorded into a playback JSON such as random UUIDs generated by methods called.

**Comment by alzimmermsft** [comment]
Closing in favor of #6671

**Pull request closed without merging**